lib/puppetserver/ca/create_action.rb

```diff
@@ -0,0 +1,268 @@
+require 'puppetserver/ca/utils'
+require 'puppetserver/ca/host'
+require 'puppetserver/ca/puppet_config'
+require 'puppetserver/utils/file_utilities'
+require 'puppetserver/utils/refactored_client'
+require 'puppetserver/utils/signing_digest'
+require 'json'
+
+module Puppetserver
+  module Ca
+    class CreateAction
+
+      include Puppetserver::Utils
+
+      # Only allow printing ascii characters, excluding /
+      VALID_CERTNAME = /\A[ -.0-~]+\Z/
+      CERTNAME_BLACKLIST = %w{--all --config}
+
+      SUMMARY = "Create a new certificate signed by the CA"
+      BANNER = <<-BANNER
+Usage:
+  puppetserver ca create [--help]
+  puppetserver ca create [--config] --certname CERTNAME[,ADDLCERTNAME]
+
+Description:
+Creates a new certificate signed by the intermediate CA
+and stores generated keys and certs on disk.
+
+To determine the target location, the default puppet.conf
+is consulted for custom values. If using a custom puppet.conf
+provide it with the --config flag
+
+Options:
+BANNER
+      def initialize(logger)
+        @logger = logger
+      end
+
+      def self.parser(parsed = {})
+        parsed['certnames'] = []
+        OptionParser.new do |opts|
+          opts.banner = BANNER
+          opts.on('--certname FOO,BAR', Array,
+               'One or more comma separated certnames') do |certs|
+            parsed['certnames'] += certs
+          end
+          opts.on('--help', 'Display this create specific help output') do |help|
+            parsed['help'] = true
+          end
+          opts.on('--config CONF', 'Path to puppet.conf') do |conf|
+            parsed['config'] = conf
+          end
+        end
+      end
+
+      def parse(args)
+        results = {}
+        parser = self.class.parser(results)
+
+        errors = Utils.parse_with_errors(parser, args)
+
+        if results['certnames'].empty?
+          errors << '    At least one certname is required to create'
+        else
+          results['certnames'].each do |certname|
+            if CERTNAME_BLACKLIST.include?(certname)
+              errors << "    Cannot manage cert named `#{certname}` from " +
+                        "the CLI, if needed use the HTTP API directly"
+            end
+
+            unless certname =~ VALID_CERTNAME
+              errors << '  Certname %{name} must not contain unprintable or non-ASCII characters'
```

Comment:
Where does `%{name}` come from here?

Reply:
Is this true?

Reply:
From some old code I forgot to change :(

Reply:
```
irb(main):001:0> VALID_CERTNAME = /\A[ -.0-~]+\Z/
=> /\A[ -.0-~]+\Z/
irb(main):002:0> "amy" =~ VALID_CERTNAME
=> 0
```
```
 "¢" =~ VALID_CERTNAME
=> nil
```
So if its valid it returns a 0, which evaluates to 0.  Else, it returns nil

Reply:
What do you mean true?  We currently check this here:
puppet/lib/puppet/ssl/base.rb:31

Reply:
I think that may be leftover from the days when we had worse unicode support. We may be able to allow other characters now. But we'll have to be careful about URL encoding for making requests using the certname. May not be worth updating right now...

Reply:
Yeah, sorry about that, I didn't mean to be so vauge with the "true" comment. 

As Maggie mentions, I don't think that that regex is actually the limit to valid certnames (according to the spec). Though at one time it may have designated the limit to certnames that Puppet could reliably handle.

Doing a bit more research it looks like any valid unicode can be in a certificate Common Name (this [s.o. answer](https://stackoverflow.com/questions/5136198/what-strings-are-allowed-in-the-common-name-attribute-in-an-x-509-certificate) is the most well written summary I've found), though we often use the CN to match DNS names which it looks like depending on the stack it often gets translated to ASCII ([wiki: internationalized domain names](https://en.wikipedia.org/wiki/Internationalized_domain_name)).

I could have swore we used to have a host `🐎.delivery.puppetlabs.net` but I don't know if our stack was converting it or what it's certname was in Puppet.

Since this is current behavior maybe it isn't worth worrying about now, I was just curious.